Patch-Source: https://github.com/CESNET/ipfixcol2/commit/9a8334732fd226d69b1b233554236041bae7fa93
--
From 9a8334732fd226d69b1b233554236041bae7fa93 Mon Sep 17 00:00:00 2001
From: Michal Sedlak <sedlakm@cesnet.cz>
Date: Tue, 19 Mar 2024 14:07:39 +0100
Subject: [PATCH] JSON output: Do not set
 PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP on systems where it doesn't
 exist

---
 src/plugins/output/json/src/File.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/plugins/output/json/src/File.cpp b/src/plugins/output/json/src/File.cpp
index bb24e3cf..6d9a7307 100644
--- a/src/plugins/output/json/src/File.cpp
+++ b/src/plugins/output/json/src/File.cpp
@@ -110,6 +110,9 @@ File::File(const struct cfg_file &cfg, ipx_ctx_t *ctx) : Output(cfg.name, ctx)
         throw std::runtime_error("(File output) Rwlockattr initialization failed!");
     }
 
+#ifdef PTHREAD_RWLOCK_WRITER_NONRECURSIVE_INITIALIZER_NP
+    // PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP avoids writer starvation
+    // It is a non-portable GNU extension only available on Linux systems
     if (pthread_rwlockattr_setkind_np(&attr, PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP) != 0) {
         if (_thread->m_calg == calg::GZIP) {
             gzclose((gzFile)_thread->file);
@@ -120,6 +123,7 @@ File::File(const struct cfg_file &cfg, ipx_ctx_t *ctx) : Output(cfg.name, ctx)
         delete _thread;
         throw std::runtime_error("(File output) Rwlockattr setkind failed!");
     }
+#endif
 
     if (pthread_rwlock_init(&_thread->rwlock, &attr) != 0) {
         if (_thread->m_calg == calg::GZIP) {
